### DISCUSSION TITLE
Next.js - slower rendering time after client side api calls components included

### DISCUSSION THREAD

**Discussion by NevilDabre** [vercel/next.js#38150] [Help]
What version of Next.js are you using?
v11.0.1

**What version of Node.js are you using?**
v14.18.2

**What browser are you using?**
Chrome

**What operating system are you using?**
macOS

**How are you deploying your application?**
cross-env next build && node index.js

**Describe the Bug**
Hi everyone, Need some help regarding HTML rendering issue happening on our production server.
Problem description -  We use Strapi (https://strapi.io/) + Next.js for our blog pages and we have some React components defined in the Strapi blog pages through the interface. On Next.js level we use Markdown to JSX (https://www.npmjs.com/package/markdown-to-jsx) package to convert the blog text to React components. The component without async/await works perfectly.  As soon as we add API calls to our Strapi components that leads to restart of our docker pod due to longer response time. 

Wrapper component with Markdown utility

```
import React from 'react';
import Markdown from 'markdown-to-jsx';
import { BlogJsonLd } from 'next-seo';
import FAQList from ‘./FAQList';
import Hero from ‘./Hero';
// With REST API Call
Import AllCountriesSection from ‘./AllCountriesSection’; 
// Tried with { ssr: false }  
// const AllCountriesSection = dynamic(() => import(‘./AllCountriesSection), { ssr: false })

const markdownOptions = {
  overrides: {
    BlogJsonLd: { component: BlogJsonLd },
    FAQPageJsonLd: { component: FAQPageJsonLd },
    Hero: { component: Hero },
    AllCountriesSection: { component: AllCountriesSection },
  }
};

function DynamicContent({ data }) {
  const { content, path, lang, slug, SEO, query } = data;

 if (!content) return null;
  let dynamicContent = content;
  return (
    <>
      <div style={{ overflowY: 'auto', overflowX: 'hidden' }}>
        <Markdown
          children={dynamicContent}
          options={markdownOptions}
        />
      </div>
    </>
  );
}

export default DynamicContent
```

Component with an API call
```
import React, { useState, useEffect } from 'react';
import { Button } from '@material-ui/core'
import Grid from '@material-ui/core/Grid';
import { connect } from 'react-redux';
import { fetchCountriesFromRates } from '../../redux/system/actions'; // REST API Call using fetch

const CurrencyListSection = ({ t, classes, countries, fetchCountries }) => {
    const [countriesCurrencyList, setCountriesCurrencyList] = useState([])
    const [loading, setLoading] = useState(true);

    useEffect(() => {
        loadTransferRates();
    }, [countries]);

    const loadTransferRates = async () => {
        setLoading(true);
        if (countries) {
            setCountriesCurrencyList(countries)
        } else {
            await fetchCountries();
        }
        setLoading(false);
    }

  if(loading) return <></>;
    
  return (
    <section className={classes.root}>
      <Grid container>
                <Grid item container xs={12} spacing={2}>
                    {countriesCurrencyList?.map(country => (
                        <Grid item xs={12} sm={6} md={4} lg={3}>
                            <Button 
                                variant="outlined"
                                onClick={() => Router.push(`/country-{country.code.toLowerCase()}`) }
                            >
                               {country.name}
                            </Button>
                        </Grid>
                    ))}
                </Grid>
      </Grid>
    </section>
  );
}


const mapStateToProps = ({ system }) => ({
    countries: system.countries,
});

const mapDispatchToProps = (dispatch) => ({
    fetchCountries: () => dispatch(fetchCountriesFromAPI()),
});

export default connect(mapStateToProps, mapDispatchToProps)(CurrencyListSection);
```

Any help or leads will be appreciated, thanks.

**Comment by icyJoseph** [comment]
So the issue is that using a component that fetches data with `useEffect`, makes it so you start to see timeouts?

Where's `fetchCountriesFromAPI` declared? I only see `fetchCountriesFromRates`.

**Comment by Babbili** [comment]
@NevilDabre since the fetching happens on a Component file and not a Page file, then you can only use client-side fetching which is slower than other [Data fetching methods](https://nextjs.org/docs/basic-features/data-fetching/overview) that you can use in NextJs Page file.
I recommend using [Client-side data fetching with SWR](https://nextjs.org/docs/basic-features/data-fetching/client-side#client-side-data-fetching-with-swr), It's faster and it handles caching, revalidation, refetching on intervals
This is how to use it, and you can see more in the [docs](https://swr.vercel.app/docs/getting-started)
```
import useSWR from 'swr'

const fetcher = (...args) => fetch(...args).then((res) => res.json())

function Countries() {
  const { data, error } = useSWR('/api/countries-data', fetcher)

  if (error) return <div>Failed to load</div>
  if (!data) return <div>Loading...</div>

  return (
    <div>
      <h1>{data.name}</h1>
    </div>
  )
}
```
#38150

**Reply by NevilDabre** [reply]
@Babbili Thanks for your reply, I will try this and update here.

**Reply by icyJoseph** [reply]
I must be missing something here, but how's this any different to the OP approach? All data fetching solutions end up using `useEffect` and such under the hood, you just don't see it, and they do a lot more like caching, deduping, etc...

There's recommendations to avoid manually rolling your own effects to fetch data, and instead use data fetching libraries, but at the end of the day, we all, libraries and production code, have access to the same APIs, we just don't reinvent the wheel every time, because we install libraries.

So what am I missing here?

**Reply by NevilDabre** [reply]
@icyJoseph I totally agree with you on this point because I need to call API only for particular component not for all components because I am using **markdown-to-jsx** to dynamically map the content at build time. Also the api data is consistently changing so need to get this data on client side only.

Next.js SSR is not useful in such scenarios. In my case, the components with API calls first render time goes up-to 30 to 60 seconds after couple of minutes deployment on production.

**Reply by icyJoseph** [reply]
Could you try to describe the bug in a timeline? Where does it fail? at build time?

> the components with API calls first render time goes up-to 30 to 60 seconds after couple of minutes deployment on production.

So these are component that, through SWR, useEffect, etc, etc, fetches data somehow, on the client, right? However, after a few minutes of deployment, each API call made by these components starts to slow down? Is that what's happening?

**Reply by NevilDabre** [reply]
Yes, I use useEffect to load the data from API server. So after deployment everything works as intended but after 1 or 2 hours all the pages on our portal takes longer to load. Basically Next.JS takes longer to build and send the response to the client. 

Originally, we used to return simple **loading icon** but it affected our SEO because the page source was coming without html content. So we removed loading icon and tried to return whole HTML content and this problem started arising on components with API calls.
 
> Timeline
1. Website is deployed using Kubernetes Docker
2. It works normally for 1 or 2 hours depends on load
3. Afterwards time for the every page load start taking longer, i.e. more than 30 seconds
4. Even Kubernetes liveness check fails as overall Next.JS server POD becomes slower.
5. Kubernetes restarts all the pods and website goes down until all pods deployed again.

If we don't render whole HTML content then there is no issue.

**Reply by icyJoseph** [reply]
How do you get data from Strapi? I guess you are using GetServerSideProps?

This sounds like a memory leak. 

Somehow we are either using up all listeners to resolve async calls, or creating a ton of DB connections to the same DB, which while keeping them all open, cannot serve any further.

When you sent an icon the data was fetched only client side.

Now that you fetch in the request, to populate as much HTML as possible, perhaps there are open handles left open in the server.

**Reply by francielelithg** [reply]
@icyJoseph thanks for your reply!! I'm currently working on this with @NevilDabre. Initially, we're loading strapi content inside a getInitialProps page call. Then we convert it using markdown-to-jsx to render the React components. So, some of these components, they have API calls inside useEffect (as he mentioned, sensitive dynamic data, that can be changed every minute).

**Reply by icyJoseph** [reply]
Aha, ok, and a consequence of this is that the server becomes slower over an hour or so, after deploy, correct?

Without a repo this is almost a guessing game for me, but it really feels like a memory leak. The API calls inside useEffect, do they go through Next.js?

**Reply by francielelithg** [reply]
That's correct, yes for both questions.
